crates/sui-types/src/dynamic_field.rs

```diff
@@ -206,28 +208,58 @@ pub fn is_dynamic_object(move_struct: &MoveStruct) -> bool {
 pub fn derive_dynamic_field_id<T>(
     parent: T,
     key_type_tag: &TypeTag,
-    key_type_layout: &MoveTypeLayout,
-    key: &move_vm_types::values::Value,
-) -> Option<ObjectID>
+    key_bytes: &[u8],
+) -> Result<ObjectID, bcs::Error>
 where
     T: Into<SuiAddress>,
 {
-    let Ok(k_tag_bytes) = bcs::to_bytes(key_type_tag) else {
-        return None;
-    };
-    let Some(k_bytes) = key.simple_serialize(key_type_layout) else {
-        return None;
-    };
+    let k_tag_bytes = bcs::to_bytes(key_type_tag)?;
 
     // hash(parent || key || key_type_tag)
     let mut hasher = Sha3_256::default();
     hasher.update(parent.into());
-    hasher.update(k_bytes.len().to_le_bytes());
-    hasher.update(k_bytes);
+    hasher.update(key_bytes.len().to_le_bytes());
+    hasher.update(key_bytes);
     hasher.update(k_tag_bytes);
     let hash = hasher.finalize();
 
     // truncate into an ObjectID and return
     // OK to access slice because Sha3_256 should never be shorter than ObjectID::LENGTH.
-    Some(ObjectID::try_from(&hash.as_ref()[0..ObjectID::LENGTH]).unwrap())
+    Ok(ObjectID::try_from(&hash.as_ref()[0..ObjectID::LENGTH]).unwrap())
+}
+
+pub trait MoveTypeTagTrait {
+    fn get_type_tag() -> TypeTag;
+}
+
+impl MoveTypeTagTrait for u64 {
+    fn get_type_tag() -> TypeTag {
+        TypeTag::U64
+    }
+}
+
+/// Given a parent object ID (e.g. a table), and a `key`, retrieve the corresponding dynamic field
+/// from the `object_store`. The key type `K` must implement `MoveTypeTagTrait` which has an associated
+/// function that returns the Move type tag. This is needed to properly derive the field object ID.
+pub fn get_dynamic_field_from_store<S, K, V>(
+    object_store: &S,
+    parent_id: ObjectID,
+    key: &K,
+) -> Result<V, SuiError>
+where
+    S: ObjectStore,
+    K: MoveTypeTagTrait + Serialize + DeserializeOwned,
+    V: Serialize + DeserializeOwned,
+{
+    let id = derive_dynamic_field_id(parent_id, &K::get_type_tag(), &bcs::to_bytes(key).unwrap())
+        .map_err(|err| SuiError::DynamicFieldReadError(err.to_string()))?;
+    let object = object_store.get_object(&id)?.ok_or_else(|| {
+        SuiError::DynamicFieldReadError("Dynamic field not found in the table".to_owned())
```

Comment:
nit, dynamic fields can exist on other types (not just tables).

```suggestion
        SuiError::DynamicFieldReadError("Dynamic field not found on parent".to_owned())
```